expression/builtin_control.go

```diff
@@ -472,7 +472,13 @@ func (c *ifFunctionClass) getFunction(ctx sessionctx.Context, args []Expression)
 	}
 	retTp := InferType4ControlFuncs(args[1].GetType(), args[2].GetType())
 	evalTps := retTp.EvalType()
-	bf := newBaseBuiltinFuncWithTp(ctx, args, evalTps, types.ETInt, evalTps, evalTps)
+	var arg0Tp types.EvalType
+	if args[0].GetType().EvalType() == types.ETReal {
+		arg0Tp = types.ETReal
+	} else {
+		arg0Tp = types.ETInt
```

Comment:
Perhaps we should consider other types of `args[0]`?

For example,
`SELECT IF(cast(0.1 as decimal(1, 1)), 1, 2);`

Reply:
i'm going to cast every type to `ETReal` and use `EvalReal` / `VecEvalReal` since didn't find a good way to handle strings like "0.1". But I could not generate `builtin_control_vec_generated.go`, what's the right step to do this?

Reply:
You can take a look at `isTrueOrFalseFunctionClass`.
If a function call is `if(XXX, YYY, YYY)`, one suggestion is to wrap a `builtinXXXIsTrueSig`  into `builtinIfYYYSig`.

```go
func (c *ifFunctionClass) getFunction() {
    firstParamFunc := isTrueOrFalseFunctionClass{}.getFunction(ctx, args[:1])
    return &builtinIfYYYSig{bf, firstParamFunc}
}
```

Then in `builtinIfYYYSig`, `firstParamFunc.evalInt` can be called to judgment the boolean result of XXX.